File: erlang/pe005.erl
-module(pe005).
-export([start/0]).

maxv() -> 20.

start() ->
  pe005(1, maxv(), 1).

pe005(I, Max, LCM) when I > Max ->
  io:format("~w~n", [LCM]);

pe005(I, Max, LCM) ->
  IsPrime = is_prime(I),
  if
    IsPrime ->
      pe005(I + 1, Max, LCM * max_base_m_lt_n(I, 1, Max));
    true -> 
      pe005(I + 1, Max, LCM)
  end.

max_base_m_lt_n(M, Last, N) when (M * Last) > N ->
  Last;

max_base_m_lt_n(M, Last, N) ->
  max_base_m_lt_n(M, (M * Last), N).

is_prime(Number) when Number == 1; Number == 0  ->
  false;

is_prime(Number) when Number == 2 ->
  true;

is_prime(Number) when Number rem 2 == 0 ->
  false;

is_prime(Number) ->
  prime_test(3, Number).

prime_test(Div, Number) when (Div * Div) > Number ->
  true;

prime_test(Div, Number) when Number rem Div == 0 ->
  false;

prime_test(Div, Number) ->
  prime_test(Div + 2, Number).

File: ocaml/pe005.ml
let maxv = 20 ;;

let divisible x y  = 
  x mod y = 0 ;;

let rec max_base_m_lt_n m last n =
  if m * last > n then
    last
  else
    max_base_m_lt_n m (m * last) n ;;

let rec prime_test n maxv =
  if (n * n) > maxv then
    true
  else
    if divisible maxv n then
      false
    else
      prime_test (n+2) maxv ;;

let is_prime n =
  match n with
  | 0           -> false
  | 1           -> false
  | 2           -> true
  | _           -> if divisible n 2 then false else prime_test 3 n ;;

let rec pe005 i max lcm =
  if i > max then
    Printf.printf "%d\n" lcm
  else
    if is_prime i then
      pe005 (i + 1) max (lcm * (max_base_m_lt_n i 1 max))
    else
      pe005 (i + 1) max lcm ;;

let main () =
  pe005 1 maxv 1 ;;

let () = main ()
